### PR TITLE
Feature: Case insensitive search on String fields

### LINKED ISSUE #43: better case insensitive search
newest prisma supports "native" case insensitive search. Would be nice if we could use that if its avialable (can be checked through introspection), instead of the "poor mans" version that is currently in place

### PR DESCRIPTION
**Pull request by sMteX**
Looking at the commit history, I'm not sure what happened and what made it so messed up (like showing commits that were supposed to be already merged etc.), but the actual diff looks ok. (It looks like commits I made earlier, but then I deleted the repo after messed up rebase and started fresh.. somehow they survived).

This PR implements case insensitive searching on String fields. Since Prisma 2.8.0 and newer, the API exposes a `mode` property on `StringFilter` and `StringNullableFilter`, that lets the user specify if he wants to filter by `default` (which corresponds to the database collation and its default sorting/filtering) or `insensitive` (this is reflected in an enum added in that version):
```gql
enum QueryMode { 
   default, 
   insensitive 
}
```

I updated the `getFilters()` function to check if the user's Prisma version supports this functionality (checks for the existence of the enum in introspection query, and if the string filters are actually using the enum), and if yes, considers all String filters as case insensitive. If the user has `<= 2.7.0` Prisma, it reverts back to the original pseudo-case insensitive implementation.

Also, with this update, I have to treat `StringFilter` and `StringNullableFilter` differently from their `Nested*` variants, because Nested variants don't have the `mode` property (but those input objects are always nested inside the original String filters which do have them, so I guess the information is propagated into its nested filters).

Because the test schema already is running on sufficient Prisma version, I had to update old tests so it expects the proper `where` object.

I also moved some logic into its own functions, so a bit of refactoring. 

As always, open to suggestion and feedback 🙂

### CODE REVIEW

**Comment by sMteX** [comment]
@macrozone 

I apologize for reminding you, but have you had any time to take a look at this? 🙂 Or is there anything you'd like me to change?

**Comment by macrozone** [comment]
> @macrozone
> 
> I apologize for reminding you, but have you had any time to take a look at this? 🙂 Or is there anything you'd like me to change?

looks fine, but commit history seems odd.

I don't want to get too picky with commits, but maybe try to rebase it:

`git rebase -i origin/master` 

shows you an "interactive" rebase. If you are using vscode you can also configure it as a git tool (https://stackoverflow.com/questions/30024353/how-to-use-visual-studio-code-as-default-editor-for-git), that helps a bit to get used to it, in particular if you are not familiar with vim.

its also possible to squash the pull request into one commit directly here in github.

basically the commit should show what has been added relative to the main branch. E.g. "fix string filter, refactor" should not appear in the log as you fixed your own change.

i guess the one commit should be:

```
feat: string fields use case insensitive mode if available
```

**Comment by sMteX** [comment]
Yes, I tried rebasing in the previous PR, but I think I messed up somehow so I decided to backup the changed files, delete entire fork and start over. How the commits survived when I forked the repo clean and started over, I have no idea. Also, doesn't the fact that it's already pushed on origin (my fork)?

**Comment by macrozone** [comment]
ok, ill fix it myself then

**Comment by sMteX** [comment]
I'm sorry for the trouble and thank you for merging. I'll try to fix it next time so it's less problems for you.

**Comment by macrozone** [comment]
no worries, i just tried the "merge and squash" feature of github, that did the job. :-)

getting used to rebasing is still something that i would recommend!

**Comment by sMteX** [comment]
Will try to educate myself, I usually tried to avoid it because I've heard of horror stories when rebase went _very_ wrong 😄

**Comment by macrozone** [comment]
> Will try to educate myself, I usually tried to avoid it because I've heard of horror stories when rebase went _very_ wrong 😄

true, it can go very wrong ;-) thats why its good to practise. 

I also thought for a long time that its not worth the effort and merging was simpler and better. But after getting use to it, i think it helps you to be more aware of the changes you do and be more clear about these changes. It requires more discipline, but it also teaches you more discipline.

The challenge i had was, that merging and rebases take a different _perspective_. Usually you rebase, when the main branch is ahead of where you started your branch. Merging in this case just takes any new change in the main branch and merges it into your current branch (and leaving a commit). With rebasing, you switch the perspective. you start with the main branch and try to re-apply your changes on that, as if the main branch wasn't ahead of your branch.
The difference gets obvious when there are some conflicts, notoriously the meaning of "their" changes and "our" changes are swapped if you compare rebasing to the main branch and merging in the main branch, that can be confusing.
The "dangerous" aspect is, that rebasing needs a force push. thats the one aspect i also do not like. force push means that you discard any changes on the remote and replace them with your changes. This can go wrong on branches where multiple people are working on, so in general _never force push on the main branch_, this is usually disabled on github and gitlab for this reason.
